pkg/syncer/pv_to_vvol_mapping.go

```diff
@@ -0,0 +1,157 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package syncer
+
+import (
+	"context"
+	"strings"
+	"errors"
+
+	cnstypes "github.com/vmware/govmomi/cns/types"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+
+	"sigs.k8s.io/vsphere-csi-driver/v2/pkg/common/prometheus"
+	"sigs.k8s.io/vsphere-csi-driver/v2/pkg/csi/service/common"
+	"sigs.k8s.io/vsphere-csi-driver/v2/pkg/csi/service/logger"
+)
+
+func csiGetPVtoVvolMapping(ctx context.Context, k8sclient clientset.Interface,
+	metadataSyncer *metadataSyncInformer) {
+	log := logger.GetLogger(ctx)
+	log.Info("csiGetPVtoVvolMapping: start")
+	// Call CNS QueryAll to get container volumes by cluster ID.
+	queryFilter := cnstypes.CnsQueryFilter{
+		ContainerClusterIds: []string{
+			metadataSyncer.configInfo.Cfg.Global.ClusterID,
+		},
+	}
+
+	querySelection := cnstypes.CnsQuerySelection{
+		Names: []string{
+			string(cnstypes.QuerySelectionNameTypeBackingObjectDetails),
+		},
+	}
+	queryAllResult, err := metadataSyncer.volumeManager.QueryAllVolume(ctx, queryFilter, querySelection)
+	if err != nil {
+		log.Errorf("csiGetPVtoVvolMapping: failed to QueryAllVolume with err=%+v", err.Error())
+		return
+	}
+
+	// Get K8s PVs in State "Bound".
+	k8sPVs, err := getBoundPVs(ctx, metadataSyncer)
+	if err != nil {
+		log.Errorf("csiGetPVtoVvolMapping: Failed to get PVs from kubernetes. Err: %+v", err)
+		return
+	}
+
+	// volumeHandleToPvcMap maps pv.Spec.CSI.VolumeHandle to the pvc object which
+	// bounded to the pv.
+	volumeHandleToPvcMap := make(volumeHandlePVCMap, len(k8sPVs))
+
+	for _, pv := range k8sPVs {
+		if pv.Spec.ClaimRef != nil && pv.Status.Phase == v1.VolumeBound {
+			pvc, err := metadataSyncer.pvcLister.PersistentVolumeClaims(
+				pv.Spec.ClaimRef.Namespace).Get(pv.Spec.ClaimRef.Name)
+			if err != nil {
+				log.Warnf("csiGetPVtoVvolMapping: Failed to get pvc for namespace %s and name %s. err=%+v",
+					pv.Spec.ClaimRef.Namespace, pv.Spec.ClaimRef.Name, err)
+				continue
+			}
+			volumeHandleToPvcMap[pv.Spec.CSI.VolumeHandle] = pvc
+			log.Debugf("csiGetPVtoVvolMapping: pvc %s/%s is backed by pv %s volumeHandle %s",
+				pvc.Namespace, pvc.Name, pv.Name, pv.Spec.CSI.VolumeHandle)
+		}
+	}
+
+	// pv to vvol Map maps vol.VolumeId.Id to backingobjectId.
+	volumeIdToBackingObjectIdMap := make(volumeIdHealthStatusMap, len(queryAllResult.Volumes))
+
+	for _, vol := range queryAllResult.Volumes {
+		volumeIdToBackingObjectIdMap[vol.VolumeId.Id] = vol.BackingObjectDetails.(*cnstypes.CnsBlockBackingDetails).BackingDiskId
+	}
+
+	for volID, pvc := range volumeHandleToPvcMap {
+		pvTovvolPair := volID + ":" + volumeIdToBackingObjectIdMap[volID]
+		updatePVtoVvolMappingStatus(ctx, k8sclient, pvc, pvTovvolPair)
```

Comment:
make the method `updatePVtoVvolMappingStatus` return an error and a boolean which indicates if the pv was updated or not, track the number of successful updates and failed updates, and log it. This way if some users were expecting the annotation and it isn't present we can trace back in the logs to check when the updates started failing.

This is a good to have, but if there is time constraint on the PR, you can skip this and work on it later.

Reply:
Add an error and boolen return to indicate whether annotation is updated.